pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTExpression.java

```diff
@@ -91,4 +94,25 @@ default boolean isCompileTimeConstant() {
         return getConstValue() != null;
     }
 
+    /**
+     * Returns the type expected by the context. This type may determine
+     * an implicit conversion of this value to that type (eg a boxing
+     * conversion, widening numeric conversion, or widening reference
+     * conversion).
+     *
+     * <p>There are many different cases.
+     * For example, in {@code arr['c']}, {@link #getTypeMirror()} would
+     * return {@code char} for the char literal, but the context type
+     * is {@code int} since it's used as an array index. Similarly,
+     * the context type of an expression in a return statement is the
+     * return type of the method, etc.
+     *
+     * <p>This returns null when that type is undefined, for example,
```

Comment:
If this method returns null, then we should remove `@NonNull` and add `@Nullable` instead?

Reply:
Ah I think I forgot to update the doc. The method never returns null. The case "undefined" is now a non-null ExprContext which will answer true to `isMissing()`. So `getConversionContextType().isMissing()` is the new `getConversionContextType() == null`. 

Also this method should probs be renamed to `getConversionContext`. It returned just a JTypeMirror before I made ExprContext public